src/engine.cc

```diff
@@ -133,33 +136,49 @@ SearchLimits EngineController::PopulateSearchLimits(int ply, bool is_black,
   if (limits.infinite || time < 0) return limits;
   int increment = std::max(int64_t(0), is_black ? params.binc : params.winc);
 
-  int movestogo = params.movestogo < 0 ? 50 : params.movestogo;
   // Fix non-standard uci command.
-  if (movestogo == 0) movestogo = 1;
+  float movestogo = params.movestogo == 0.0f ? 1.0f : params.movestogo;
 
   // How to scale moves time.
   float slowmover = options_.Get<float>(kSlowMoverStr);
-  int64_t move_overhead = options_.Get<int>(kMoveOverheadStr);
-  float time_curve_peak = options_.Get<float>(kTimeCurvePeak);
-  float time_curve_left_width = options_.Get<float>(kTimeCurveLeftWidth);
-  float time_curve_right_width = options_.Get<float>(kTimeCurveRightWidth);
+  int move_overhead = options_.Get<int>(kMoveOverheadStr);
+
+  float time_curve_midpoint = options_.Get<float>(kTimeCurveMidpoint);
+  float time_curve_scale = options_.Get<float>(kTimeCurveScale);
+  float time_curve_shape = options_.Get<float>(kTimeCurveShape);
+
+  // Use a ply-based heuristic to guess how many moves ahead are worth planning
+  // time for.
+  float guessed_movestogo = 0.0f;
+  // Compute first weight separately as it will be used to normalise.
+  float this_move_prob = ComputePlyWeight(ply, time_curve_midpoint,
+                                          time_curve_scale, time_curve_shape);
+
+  // Integrate over a large range of plies to approximate integrating to
+  // infinity.
+  for (int i = 1; i < ply + 300; ++i) {
+    guessed_movestogo += ComputePlyWeight(ply + 2 * i, time_curve_midpoint,
+                                          time_curve_scale, time_curve_shape);
+  }
```

Comment:
Surely you can compute the analytic integral, or a good approximation to it, and then compute that directly instead of this tiresome loop?

Reply:
Yes, possibly. I'll investigate.

Reply:
This seems to calculate the mean residual time. A web search found a solution in <http://www.lifescienceglobal.com/pms/index.php/ijsmr/article/download/2691/1579> equation (3). The same reference also gives in equation (5) a simpler formula for the median residual time.